core/src/main/java/com/netflix/iceberg/util/ByteBuffers.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.netflix.iceberg.util;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+public class ByteBuffers {
+
+  public static byte[] copy(ByteBuffer buffer) {
```

Comment:
This method doesn't actually provide a defensive copy in all cases, so it shouldn't be a helper that claims to make a copy. A better name is `toByteArray` because it will use the underlying byte array if it exists and has an appropriate offset and length.

I think ByteBuffers should provide both `copy` and `toByteArray` and `SerializableByteBufferMap` should continue to use `toByteArray`.

Reply:
In most cases how worried are we about mutation of the returned array such that we warrant a defensive copy? Think we want to avoid defensive copy wherever we can get away with it because of performance concerns.

Reply:
We definitely want to avoid defensive copies. My point here is more about the promise the name "copy" makes. It should not claim to copy and not actually copy (my bad).

If we don't need a copy when extracting the value from the ByteBuffer, that would be ideal. For data files in ScanTasks, we should not need to copy because a defensive copy is made in the planning process. That's required because records are reused in deserialization to avoid instantiation and GC overhead. Once past the scan filters, they have to be copied.

Reply:
Ok, what I've done then is make an explicit `copy` method and am using that in the copy constructors and copy builders for data files. But I'd expect that we shouldn't be calling such copy constructors outside of contexts where we actually want a deep copy of the data.